serviceWorker.js

```diff
@@ -0,0 +1,108 @@
+	const newHeaders = new Headers(r.headers);
+
+	// COEP & COOP for cross-origin-isolation
+	newHeaders.set("Cross-Origin-Embedder-Policy", "require-corp");
+	newHeaders.set("Cross-Origin-Opener-Policy", "same-origin");
+
+	newHeaders.set("Cross-Origin-Resource-Policy", "cross-origin");
+
+	/** 
+	 * This is a hacky fix, due to cheerpOS relying on the response.url being set to the resolved url (after a redirect)
+	 * when we construct a new response, it will not have the correct url set, and the serviceworker respondwith method will set the url
+	 * to the value of event.request.url; see https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent/respondWith.
+	 * In case of a redirect, we therefore set the location url to the resolved response url, and set statuscode to 301 in our new 
+	 * Response object, this then bounces the request back to the serviceworker from Cheerpos, with the request.url being set to the resolved url, therefore responseWith will set it properly in our new response.
+	 */
+	if (r.redirected === true)
+		newHeaders.set("location", r.url);
+
+	// In case of a redirection, we set the status to 301, and body to null, in order to not transfer too much data needlessly
+	const moddedResponse = new Response(r.redirected === true ? null : r.body, {
+		headers: newHeaders,
+		status: r.redirected === true ? 301 : r.status,
+		statusText: r.statusText,
+	});
+	return moddedResponse;
+}
+
+function serviceWorkerInit() {
+	// Init the service worker.
+	self.addEventListener("install", () => self.skipWaiting());
+	self.addEventListener("activate", e => e.waitUntil(self.clients.claim()));
+
+	// Listen for fetch requests and call handleFetch function.
+	self.addEventListener("fetch", function (e) {
+		try {
+			e.respondWith(handleFetch(e.request));
+		} catch (err) {
+			verbose && console.log("Serviceworker respondWith failed with ERROR:" + err);
+		}
+	});
+}
+
+function doRegister() {
+	navigator.serviceWorker.register(window.document.currentScript.src)
+		.then((registration) => {
+			verbose && console.log("Service Worker registered", registration.scope);
+
+			// Reload the page to make use of an updated service worker
+			registration.addEventListener("updatefound", () => {
+				verbose && console.log("Reloading page to use the updated Service Worker.");
+				try {
+					window.location.reload();
+				} catch (err) {
+					verbose && console.log("Serviceworker failed reloading the page. ERROR:" + err);
+				};
+			});
+
+			// If the registration is active, but it's not controlling the page, reload the page to have it take control
+			if (registration.active && !navigator.serviceWorker.controller) {
+				verbose && console.log("Reloading the page to transfer control to the Service Worker.");
+				try {
+					window.location.reload();
+				} catch (err) {
+					verbose && console.log("Service Worker failed reloading the page. ERROR:" + err);
```

Comment:
Can "reload" even throw? Also, print always the message, it would be fatal

Reply:
Yes, reload can throw : https://developer.mozilla.org/en-US/docs/Web/API/Location/reload

"The reload may be blocked and a SECURITY_ERROR [DOMException](https://developer.mozilla.org/en-US/docs/Web/API/DOMException) thrown."